[Xcode] `build-dumprendertree` and other scripts do not support workspace builds

emw-apple opened the pull request:
#### b03221eccffc484cc8ba7d889fd03f595f5449e6
<pre>
[Xcode] `build-dumprendertree` and other scripts do not support workspace builds
<a href="https://bugs.webkit.org/show_bug.cgi?id=243233">https://bugs.webkit.org/show_bug.cgi?id=243233</a>
rdar://97615725

Reviewed by Alexey Proskuryakov.

Adjust the build-* family of scripts to call buildXcodeScheme, instead
of building individual projects. They now build a scheme in the WebKit
workspace. These schemes (DumpRenderTree, TestWebKitAPI, etc.) do not
build their implicit dependencies, so it&apos;s still possible to build these
tools without implicitly building the entire WebKit stack.

* Source/WebCore/WebCore.xcodeproj/xcshareddata/xcschemes/WebCore.xcscheme:
* Tools/DumpRenderTree/DumpRenderTree.xcodeproj/xcshareddata/xcschemes/DumpRenderTree.xcscheme:
* Tools/Scripts/build-api-tests:
* Tools/Scripts/build-dumprendertree:
* Tools/Scripts/build-imagediff: The only exception: it&apos;s not part of
  any workspace and never has been, so call
  overrideConfiguredXcodeWorkspace to ensure we never try to build it in
  a workspace.
* Tools/Scripts/build-lldbwebkittester:
(buildProjectOrDie): Deleted.
* Tools/Scripts/build-webkit:
* Tools/Scripts/build-webkittestrunner:
* Tools/Scripts/webkitdirs.pm:
(XcodeOptions): Instead of the previous behavior, where webkitdirs would
not build in a workspace _at all_ if one hadn&apos;t been set, now fall back
to WebKit.xcworkspace by default. Scripts can still opt-out of building
in a workspace by calling overrideConfiguredXcodeWorkspace(&quot;&quot;).
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWTF.xcscheme: Added, builds TestWTF and gtest.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme: Also build gtest in addition to TestWebKitAPI.
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/xcshareddata/xcschemes/WebKitTestRunner.xcscheme:
* Tools/lldb/lldbWebKitTester/lldbWebKitTester.xcodeproj/xcshareddata/xcschemes/lldbWebKitTester.xcscheme: Added.

Canonical link: <a href="https://commits.webkit.org/252857@main">https://commits.webkit.org/252857@main</a>
</pre>

emw-apple commented on Tools/Scripts/build-webkit:
```diff
@@ -226,11 +226,9 @@ if ((isAppleWebKit() || isWinCairo() || isPlayStation() || isFTW()) && !-d "WebK
     if (!$useWorkspace) {
+        # Ignore any workspace set by set-webkit-configuration
         overrideConfiguredXcodeWorkspace("");
-    } else {
-        overrideConfiguredXcodeWorkspace("WebKit.xcworkspace");
```
This represents a slight behavior change: Instead of _always_ using WebKit.xcworkspace, `build-webkit` may now use other configured workspaces. For internal engineers, this means that `build-webkit` will share incremental build data with other workspaces.

aproskuryakov replied:
Not sure if this is a scenario that is or should be supported. This change is a no-op for open source builds AFAICT, so it's OK.

emw-apple replied:
I know we document `build-webkit` as the preferred workflow for open source builds exclusively, but I think that using the configured workspace will make it do the expected thing for engineers running it internally. It also aligns `build-webkit`'s behavior with the other `build-*` scripts here.

You are correct that it's a no-op for open source builds.

aproskuryakov commented on Tools/Scripts/build-api-tests:
```diff
@@ -63,15 +63,8 @@ chdirWebKit();
-        push @options, ("-target", "TestWTF");
-    }
-    $result = buildXCodeProject("TestWebKitAPI", $clean, @options, @ARGV);
+    my $scheme = $wtfOnly ? "TestWTF" : "TestWebKitAPI";
```
Does this still build gtest? Looking at the xcscheme, I think that the answer is yes, but want to confirm.

emw-apple replied:
Yes!

emw-apple committed:
[Xcode] `build-dumprendertree` and other scripts do not support workspace builds
https://bugs.webkit.org/show_bug.cgi?id=243233
rdar://97615725

Reviewed by Alexey Proskuryakov.

Adjust the build-* family of scripts to call buildXcodeScheme, instead
of building individual projects. They now build a scheme in the WebKit
workspace. These schemes (DumpRenderTree, TestWebKitAPI, etc.) do not
build their implicit dependencies, so it's still possible to build these
tools without implicitly building the entire WebKit stack.

* Source/WebCore/WebCore.xcodeproj/xcshareddata/xcschemes/WebCore.xcscheme:
* Tools/DumpRenderTree/DumpRenderTree.xcodeproj/xcshareddata/xcschemes/DumpRenderTree.xcscheme:
* Tools/Scripts/build-api-tests:
* Tools/Scripts/build-dumprendertree:
* Tools/Scripts/build-imagediff: The only exception: it's not part of
  any workspace and never has been, so call
  overrideConfiguredXcodeWorkspace to ensure we never try to build it in
  a workspace.
* Tools/Scripts/build-lldbwebkittester:
(buildProjectOrDie): Deleted.
* Tools/Scripts/build-webkit:
* Tools/Scripts/build-webkittestrunner:
* Tools/Scripts/webkitdirs.pm:
(XcodeOptions): Instead of the previous behavior, where webkitdirs would
not build in a workspace _at all_ if one hadn't been set, now fall back
to WebKit.xcworkspace by default. Scripts can still opt-out of building
in a workspace by calling overrideConfiguredXcodeWorkspace("").
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWTF.xcscheme: Added, builds TestWTF and gtest.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/xcshareddata/xcschemes/TestWebKitAPI.xcscheme: Also build gtest in addition to TestWebKitAPI.
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/xcshareddata/xcschemes/WebKitTestRunner.xcscheme:
* Tools/lldb/lldbWebKitTester/lldbWebKitTester.xcodeproj/xcshareddata/xcschemes/lldbWebKitTester.xcscheme: Added.

Canonical link: https://commits.webkit.org/252857@main
